File: LeetCode/Algorithms/Easy/PascalsTriangle/PascalsTriangle.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        result = [[1]]
        for i in range(1, numRows):
            temp1 = result[-1] + [0]
            temp2 = [0] + result[-1]
            result.append([temp1[i] + temp2[i] for i in range(len(temp1))])
        return result[:numRows]
            
        
# Time Complexity - O(n**2)\
# Space Complexity - O(n)
        
"""
explanation: Any row can be constructed using the offset sum of the previous row. Example:

    1 3 3 1 0 
 +  0 1 3 3 1
 =  1 4 6 4 1
 """

File: LeetCode/Algorithms/Easy/PascalsTriangle/PascalsTriangle.java
import java.util.ArrayList;
import java.util.List;

public class PascalTriangle {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> result = new ArrayList();
        if(numRows == 0)
            return result;

        List<Integer> first_row = new ArrayList<>();
        first_row.add(1);
        result.add(first_row);

        for(int i = 1; i < numRows; i++){
            List<Integer> prev_row = result.get(i - 1);
            List<Integer> row = new ArrayList<>();
            row.add(1);
            for(int j = 1; j < i; j++){
                row.add(prev_row.get(j - 1) + prev_row.get(j));
            }
            row.add(1);
            result.add(row);
        }
        return result;
    }
}
